### DISCUSSION TITLE
Weird behavior with device creation script

### DISCUSSION THREAD

**Discussion by llamafilm** [netbox-community/netbox#15718] [General]
I have a custom script to create devices.  I noticed a problem where new devices created this way do not show any power ports, outlets, and interfaces in the GUI, although they do exist, and the API shows counts of 0, for example `/api/dcim/devices/8178/` shows:

```
"power_port_count": 0,
"power_outlet_count": 0,
"interface_count": 0,
```

The script runs like this (simplified)
```python
# create device
dev = Device(site=xxx, status='active'...)
dev.full_clean()
dev.save()

# create IP address
ipaddr = IPAddress(address=data["primary_ip"])
ipaddr.full_clean()
ipaddr.save()

# assign primary IP to device
iface = Interface.objects.get(device_id=dev.id, name='Ethernet1')
iface.ip_addresses.set([ipaddr])
ipaddr = IPAddress.objects.get(id=ipaddr.id)
dev.primary_ip4 = ipaddr
dev.full_clean()
dev.save()                    
```

It turns out, I have to run `dev = Device.objects.get(id=dev.id)` after saving the device, and that fixes the issue.

**Comment by kkthxbye-code** [comment]
Do `dev.refresh_from_db()` like shown here:

```python
# create device
dev = Device(site=xxx, status='active'...)
dev.full_clean()
dev.save()

# create IP address
ipaddr = IPAddress(address=data["primary_ip"])
ipaddr.full_clean()
ipaddr.save()

# assign primary IP to device
iface = Interface.objects.get(device_id=dev.id, name='Ethernet1')
iface.ip_addresses.set([ipaddr])
ipaddr = IPAddress.objects.get(id=ipaddr.id)

dev.refresh_from_db()
dev.primary_ip4 = ipaddr
dev.full_clean()
dev.save()
```

Netbox added a half-baked caching implementation of component counts recently. This means that whenever you save a device, a signal is executed afterwards for calculating the count, meaning that the device object will always have to be reloaded after save to make sure you are not overwriting any post-save changes.

**Comment by rizlas** [comment]
Check this https://github.com/netbox-community/customizations/issues/100#issuecomment-2026022272

**Reply by candlerb** [reply]
Yes, you can fix the cached counts when they get broken, but better not to break them in the first place.

This means that if you have an instance of an object which contains component counts (e.g. a device), then you need to do `device.refresh_from_db()` before updating and saving.

**Reply by rizlas** [reply]
Thanks, will update scripts with refresh from db function

**Reply by rizlas** [reply]
Can you elaborate more on how to use refresh_from_db??
When the bug will occur?

For example if I create an Interface, after saving the interface I must also "resave" the associated device?

eg:

```python
iface = Interface(....)
iface.device = dev
iface.full_clean()
iface.save()

dev.refresh_from_db()
dev.full_clean()
dev.save()
```

**Reply by candlerb** [reply]
You don't have to to include the last three lines at all, if you have no need to touch "dev" further.

The problem is this: creating the interface and adding it to a device updates the cached interface count on the device row *in the database*, but doesn't update the local copy you have stored in the "dev" variable.

Therefore, if you wanted to modify and re-save the device, within the same script, you'd need to refresh your copy in RAM:

```
dev.refresh_from_db()    # updates all local attributes, including the cached component counts
dev.description = "New description"
dev.full_clean()
dev.save()
```

... since `dev.save()` saves *all* attributes on the device, not just those you've explicitly modified. Welcome to the world of ORMs.

**Reply by rizlas** [reply]
That make sense and in a part of my code where i was modifying the device adding refresh_from_db fixed the count. It does not work in another part where, for example, I'm only adding a rear/console ports to the device without any further modification to the device.

Here a part of the code:

```python
def _setup_default_ports(self, dev_obj):
    try:
        device_rp = RearPort.objects.get(
            device=dev_obj,
            name__startswith=f"Backplane {dev_obj.name}",
        )
    except RearPort.DoesNotExist:
        device_rp = RearPort(
            device=dev_obj,
            description=f"Backplane {dev_obj.name}",
            type="other",
            name=f"Backplane {dev_obj.name}",
        )
    except Exception as e:
        self.log_warning(f"Error getting RearPort for {dev_obj.name} {e}")
        return (None, -1)

    changed = False
    self.obj_snapshot(device_rp)

    if not device_rp.description:
        device_rp.description = f"Backplane {dev_obj.name}"
        changed = True
    if not device_rp.type:
        device_rp.type = "other"
        changed = True
    if device_rp.positions != 256:  # noqa: PLR2004
        device_rp.positions = 256
        changed = True
    if changed:
        device_rp.full_clean()
        device_rp.save()

    cp, _ = ConsolePort.objects.get_or_create(device=dev_obj, name="Console - RE0")
```

**Reply by candlerb** [reply]
Note that

```
        device_rp = RearPort(
            device=dev_obj,
            description=f"Backplane {dev_obj.name}",
            type="other",
            name=f"Backplane {dev_obj.name}",
        )
```

does not create a RearPort in the database until you save it. So I think you need to set changed=True here (and move changed=False earlier, obviously; you also only need to snapshot the device if it already exists and you read it from the database)

The console port, however, should be fine, as long as you don't call `dev_obj.save()` after this point.

**Reply by rizlas** [reply]
Of course I'm saving them. Also, weirdly enough neither console port is shown :-| I'm completely clueless.
I've also tried to leave only console port creation and remove the rest. Only interfaces tab appears.

**Reply by rizlas** [reply]
Ok find it.. 

```python
def _setup_default_ports(self, dev_obj):
```

dev_obj is a Device instance (get or created) at the start of the script. 

adding dev_obj.refresh_from_db() and then the snapshot fixed the problem.

**Reply by rizlas** [reply]
Thanks for the explanation and help :)
